spec/consensus/light-client.md

```diff
@@ -1,318 +1,577 @@
+Note that in case `VerifyAndUpdateSingle` returns without an error (untrusted header
+is successfully verified) then we have a guarantee that the transition of the trust
+from `trustedState` to `newTrustedState` happened during the trusted period of
+`trustedState.SignedHeader.Header`.
 
-- the lite client communicates with one full node
-- the lite client locally stores all the signed headers it obtained (trusted or not). In the pseudo code below we write _Store(header)_ for this.
-- If _Bisection_ returns _false_, then the lite client has seen a forged header.
+**VerifyAndUpdateBisection.** The function `VerifyAndUpdateBisection` implements
+recursive logic for checking if it is possible building trust
+relationship between trustedState and untrusted header at the given height over
+finite set of (downloaded and verified) headers.
 
-  - However, it does not know which header(s) is/are the problematic one(s).
-  - In this case, the lite client can submit (some of) the headers it has seen as evidence. As the lite client communicates with one full node only when executing Bisection, there are two cases
-    - the full node is faulty
-    - the full node is correct and there was a fork in Tendermint consensus. Header _h1_ is from a different branch than the one taken by the full node. This case is not focus of this document, but will be treated in the document on fork accountability.
+```go
+func VerifyAndUpdateBisection(trustedState TrustedState,
+                              untrustedHeight int64,
+                              trustThreshold TrustThreshold,
+                              trustingPeriod Duration,
+                              clockDrift Duration,
+                              now Time) (error, TrustedState) {
+
+    trustedHeader = trustedState.SignedHeader.Header
+    assert trustedHeader.Height < untrustedHeight
+
+    untrustedSh, error := Commit(untrustedHeight)
+    if error != nil return (error, trustedState)
+
+    untrustedHeader = untrustedSh.Header
+    assert trustedHeader.Time < untrustedHeader.Time
+
+    // note that we pass now during the recursive calls. This is fine as
+    // all other untrusted headers we download during recursion will be
+    // for a smaller heights, and therefore should happen before.
+    assert untrustedHeader.Time < now + clockDrift
+
+    untrustedVs, error := Validators(untrustedHeight)
+    if error != nil return (error, trustedState)
+
+    untrustedNextVs, error := Validators(untrustedHeight + 1)
+    if error != nil return (error, trustedState)
+
+    error = verifySingle(
+             trustedState,
+             untrustedSh,
+             untrustedVs,
+             untrustedNextVs,
+             trustThreshold)
+
+    if fatalError(error) return (error, trustedState)
+
+    if error == nil {
+        // the untrusted header is now trusted.
+        newTrustedState = TrustedState(untrustedSh, untrustedNextVs)
+        return (nil, newTrustedState)
+    }
 
```

Comment:
Should we be explicit with what error gets us here ?

Reply:
As I guess we want errors to be part of the specifications, we should be more precise.